src/main/scala/tech/cryptonomic/conseil/tezos/michelson/contracts/TokenLedgers.scala

```diff
@@ -0,0 +1,185 @@
+   *   balance information from it
+   */
+  private case class LedgerToolbox(
+      mapId: BigMapId,
+      balanceReader: BalanceReader
+  )
+
+  /** For each specific contract available we store a few
+    * relevant bits of data useful to extract information
+    * related to that specific contract shape
+    *
+    * We're assuming here that the risk of collision on contract addresses representing
+    * tokens, even on different tezos networks, is actually zero.
+    */
+  private val ledgers: Map[ContractId, LedgerToolbox] = Map(
+    ContractId("KT1RmDuQ6LaTFfLrVtKNcBJkMgvnopEATJux") -> LedgerToolbox(
+          BigMapId(1718),
+          KT1RmDuQ6LaTFfLrVtKNcBJkMgvnopEATJux_Balance_Read
+        )
+  )
+
+  //this code needs to take into account an input of available account ids and check each for a matching balance value
+  private lazy val KT1RmDuQ6LaTFfLrVtKNcBJkMgvnopEATJux_Balance_Read: BalanceReader = update =>
+    for {
+      code <- update.value
+      balance <- MichelineOps.parseBalanceFromMap(code)
+    } yield update.key_hash -> balance
+
+  /** Does the Id reference a known token smart contract? */
+  def isKnownToken(token: ContractId): Boolean = ledgers.contains(token)
+
+  /** Extracts any available balance changes for a given reference contract representing a known token ledger
+    *
+    * @param token the id for a smart contract
+    * @param diff the big map changes found in a transaction
+    * @return a possible pair of an account and its new balance for a token associated to the passed-in contract
+    */
+  def readBalance(token: ContractId)(diff: Contract.BigMapUpdate): Option[BalanceUpdate] = diff match {
+    //we're looking for known token ledgers based on the contract id and the specific map identified by a diff
+    case update @ Contract.BigMapUpdate("update", _, _, Decimal(mapId), _) =>
+      for {
+        LedgerToolbox(BigMapId(id), customReadBalance) <- ledgers.get(token)
+        if mapId == id
+        balanceChange <- customReadBalance(update)
+      } yield balanceChange
+    case _ =>
+      None
+  }
+
+  /* Given the key hash stored in the big map and an account, checks if the
+   * two match, after encoding the account address first.
+   *
+   * @param keyHash a b58check script hash, corresponding to a map key in bytes
+   * @param accountReference the address we expect to match
+   * @return true if encoding the address and the key gives the same result
+   */
+  def hashCheck(keyHash: ScriptId)(accountReference: AccountId): Boolean = {
+    val check = Codecs.computeKeyHash(accountReference)
+
+    check.failed.foreach(
+      err =>
+        logger.error(
+          "I couldn't check big maps token updates. Failure to check hash encoding of the expected account against the map key",
+          err
+        )
+    )
+
+    check.exists(_ == keyHash.value)
+  }
+
+  /** Defines enc-dec operation used for token big maps */
+  object Codecs {
```

Comment:
Were you not able to use the existing functions in `CryptoUtil` with a little modification? These functions are generally useful so they should reside in a common utility package.

Reply:
Damn, I really missed that... 💩

Reply:
Re-used existing code, and "extended" it a bit.

I kept a couple of methods still inside the TokenContract, because they seem to me to be more tied to tezos specific logic than general purpose cryptographic functions.

Please let me know if it's fine now.